chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+4. Items referring to the spans starting at the same position have to occur in the same order before and after the post-processing.
+   Concretely, this ensures that the tree-induced hierarchy is maintained, more precise elements follow broader ones
+
+This first two properties are an implication of the preceding processes.
+All elements are derived from AST nodes, which are parsed from files retaining their position.
+Nodes that are generated by the runtime before being passed to the language server are either ignored or annotated with synthetic positions that are known to be in the bounds of the file and meet the second requirement.
+For all other nodes the second requirement is automatically fulfilled by the grammar of the Nickel language.
+The last requirement is achieved by using a stable sort during the post-processing.
+
+The algorithm used is listed in [@lst:nls-resolve-at].
+Given a concrete position, that is a `FileId` and `ByteIndex` in that file, a binary search is used to find the *last* element that *starts* at the given position.
+According to the aforementioned preconditions an element found there is equivalent to being the most specific element starting at this position.
+In the more frequent case that no element starting at the provided position is found, the search instead yields an index which can be used as a starting point to iterate the linearization *backwards* to find an item with the shortest span containing the queried position.
+Due to the third requirement, this reverse iteration can be aborted once an item's span ends before the query.
+If the search has to be aborted, the query does not have a corresponding `LinearizationItem`.
+
+```{.rust #lst:nls-resolve-at caption="Resolution of item at given position"}
+impl Completed {
+  pub fn item_at(
+    &self,
+    locator: &(FileId, ByteIndex),
+  ) -> Option<&LinearizationItem<Resolved>> {
+    let (file_id, start) = locator;
+    let linearization = &self.linearization;
+    let item = match linearization
+      .binary_search_by_key(
+        locator,
+        |item| (item.pos.src_id, item.pos.start))
+      {
+        // Found item(s) starting at `locator`
+        // search for most precise element
+        Ok(index) => linearization[index..]
+          .iter()
+          .take_while(|item| (item.pos.src_id, item.pos.start) == locator)
+          .last(),
+        // No perfect match found
+        // iterate back finding the first wrapping linearization item
+        Err(index) => {
+          linearization[..index].iter().rfind(|item| {
+            // Return the first (innermost) matching item
+            file_id == &item.pos.src_id
+            && start > &item.pos.start
+            && start < &item.pos.end
+          })
+        }
+    };
+    item
+  }
+}
+```
+
+#### Resolving by ID
+
+During the building process item IDs are equal to their index in the underlying List which allows for efficient access by ID.
```

Comment:
```suggestion
During the building process item IDs are equal to their index in the underlying array which allows for efficient access by ID.
```
Or `Vec`. It's a nitpick but list is a bit troubling because it reads like it is a linked list, which it surely shouldn't, given the complexity characteristics of lists.